docs/guide/api-vite-environment.md

```diff
@@ -804,7 +804,74 @@ await createServer({
 
 :::
 
-`RemoteRunnerTransport` and `RemoteEnvironmentTransport` are meant to be used together. If you don't use either of them, then you can define your own function to communicate between the runner and the server.
+`RemoteRunnerTransport` and `RemoteEnvironmentTransport` are meant to be used together, but you don't have to use them at all. You can define your own function to communicate between the runner and the server. For example, if you connect to the environment via an HTTP request, you can call `fetch().json()` in `fetchModule` function:
+
+```ts
+import { ESModulesEvaluator, ModuleRunner } from 'vite/module-runner'
+
+const runner = new ModuleRunner(
+  {
+    root: fileURLToPath(new URL('./', import.meta.url)),
+    transport: {
+      async fetchModule(id, importer) {
+        const response = await fetch(
+          `http://my-vite-server/fetch?id=${id}&importer=${importer}`,
+        )
+        return response.json()
+      },
+    },
+  },
+  new ESModulesEvaluator(),
+)
+
+await runner.import('/entry.js')
+```
+
+::: warning Acessing Module on the Server
+We do not want to encourage communication between the server and the runner. One of the problems that were exposed with `vite.ssrLoadModule` is over-reliance on the server state inside the processed modules. This makes it harder to implement runtime-agnostic SSR since there might be no direct access to the server APIs that are available in Node.js.
+
+Instead, we recommend using virtual modules to import the state and process it inside the loaded environment module:
+
+```ts
+import { runner } from './ssr-module-runner.js'
+import { processRoutes } from './routes.js'
+
+const { routes } = await runner.import('virtual:ssr-routes')
+processRoutes(routes)
+```
```

Comment:
I'm not sure what code was the anti-pattern from this sample code. If you have something concrete in your mind, can you add that too?

Reply:
Added an example: https://github.com/vitejs/vite/pull/17660/commits/259cb372e1854807d2417a236895380c924ccbbc

Reply:
I'm not sure if this example explains the best. What I think is an anti-pattern (i.e. not remote runner compatible) is how "runtime" code accesses resources on dev server.

One simplest example is `index.html`. If it weren't sandboxed environment, `entry-server.js` (this is what I mean by runtime code) can simply access it by `fs.readFile("index.html")`. But this can be easily make it remote-runner-compatible by using `import("/index.html?raw")` instead.

Then additionally, If users need to make use of `transformIndexHtml`, then previously it's possible to do `configureServer(server) { globalThis.__vite_server = server }` in the plugin then runtime code to access `__vite_server.transformIndexHtml`. To make this remote-runner-compatible, users need to make this part as virtual module instead of simple `?raw` import.

In this aspect, I think Remix is the most "remote runner compatible" framework and thus Frandiox (hydrogen), Dario, and I are able to run Remix on Cloudflare workerd. 

Actually, there's one thing Remix relies on `globalThis` to let server runtime easily collect css, but I was thinking that this should be rewritten as virtual module like `virtual:remix-critical-css/(path)`. Currently, I think Frandiox dealt with this by proxying `globalThis.__remix_devServerHooks` between two worlds by https://github.com/Shopify/hydrogen/blob/7d0bfe8cdf7dc333fba3c16c37bd3497e996bfb3/packages/hydrogen/src/vite/plugin.ts#L136-L160 and I also did something similar https://github.com/hi-ogawa/vite-plugins/blob/d1a1e2ac81260e9c426b33ee3bd318e6454fe913/packages/vite-node-miniflare/examples/remix/vite.config.ts#L32-L38
This could be an anti-pattern example, but probably it's rather too specific.

Reply:
My example shows that the server state is not tied to Vite, but is still stored on the server (by loading a config, for example) - this is quite easy to show as an example because it's easy to migrate. The problems I see with `index.html` example:
- you said that not every framework needs this method, so the example can be rather obscure if you don't know what it is (I was very confused by this function for a long time until I figured out why it's needed)
- the example with `index.html` doesn't call `ssrLoadModule` which is crucial to showcase in the module runner example
- I don't think it showcases the coupling with the server enough (the current example also suffers from it)

Maybe have an example with `.moduleGraph`? It would also be transformed into a virtual module, I guess